### ISSUE TITLE
question: Scoped attributes for keyed components inside loops

### ISSUE THREAD

**Issue by alexdulin** [marko-js/marko#888]
### Details
Prior to marko v4.5.0, [keyed elements in loops](http://markojs.com/docs/components/#referencing-nested-html-elements-and-components) would not work with `for:key` on corresponding elements at the index in the loop, i.e.:

```marko
$ var options =  [
  {label: 'foo', value: 'bar'},
  {label: 'bim', value: 'baz'}
];

<div.foo-div for(option in options)>
  <input.foo-input type="checkbox" value="${option.value}" key="fooOption[]"/>
  <label.foo-label for:key="fooOption[]">
    ${option.label}
  </label>
</div>
```

would break because the associated value of the index loop would not be applied to the `for` attribute of the label. This could be fixed via some hackery by using the component ID and the index value of the loop. Example:

```marko
<div.foo-div for(option in options | status-var=fooLoop)>
  <input.foo-input type="checkbox" value="${option.value}" key="fooOption[]"/>
  <label.foo-label for="${component.id}-fooOption[${fooLoop.getIndex()}]">
    ${option.label}
  </label>
</div>
```

However, without [component.id](https://github.com/marko-js/marko/blob/master/CHANGELOG.md#450) this is no longer possible, even if you switch out the `for:key` with `for:scoped="fooOption[]"` and `id:scoped="fooOption[]" key="fooOption[]"`. 

**TLDR:** How can you maintain loop index position for scoped/keyed attributes?

### Expected Behavior
Scoped/keyed attributes should be assigned the same value as the component element key they are being associated with.

### Actual Behavior
The scoped/keyed attributes do not, and end up having no variable interpolation done (the attributes remain as `fooOption[]`).

### Your Environment
Chrome 61, Node.js 8.7.0

### Stack Trace
```
Render error. Exception: TypeError: this.___._j_ is not a function
    at ComponentDef.elId (node_modules/marko/dist/components/ComponentDef.js:65:33)
    at views/components/selectize-input/components/selectize-dropdown/index.marko.js:45:36
    at forEachHelper (node_modules/marko/dist/runtime/helpers.js:89:17)
    at render (views/components/selectize-input/components/selectize-dropdown/index.marko.js:25:3)
    at renderer (node_modules/marko/dist/components/renderer.js:173:9)
    at wrappedRenderer (node_modules/marko/dist/runtime/helpers.js:107:13)
    at render (views/components/selectize-input/index.marko.js:97:3)
    at renderer (node_modules/marko/dist/components/renderer.js:173:9)
    at wrappedRenderer (node_modules/marko/dist/runtime/helpers.js:107:13)
    at render (views/components/build-sections/index.marko.js:38:3)
    at renderer (node_modules/marko/dist/components/renderer.js:173:9)
    at wrappedRenderer (node_modules/marko/dist/runtime/helpers.js:107:13)
    at render (views/pages/build.js:55:3)
    at renderer (node_modules/marko/dist/components/renderer.js:173:9)
    at safeRender (node_modules/marko/dist/runtime/renderable.js:6:9)
    at Template.render (node_modules/marko/dist/runtime/renderable.js:132:20)
    at ServerResponse.renderPage (lib/renderer/renderer.js:152:33)
    at routes/build.js:144:7
    at <anonymous>
```

**Comment by patrick-steele-idem** [comment]
I think what you are looking for is the following:

```marko
class {
  getInputs() {
    var inputEls = this.getEls('inputs');
  }
}

$ var options =  [
  {label: 'foo', value: 'bar'},
  {label: 'bim', value: 'baz'}
];

<div.foo-div for(i, option in options)>
  
  <input.foo-input type="checkbox" value="${option.value}"
    key="inputs[]"
    id:scoped="input-${i}"/>
    
  <label.foo-label for:scoped="input-${i}">
    ${option.label}
  </label>
  
</div>
```

More details:

- Use the [`:scoped` modifier on the attribute](http://markojs.com/docs/components/#codescopedcode) to create a globally unique ID from a locally unique ID (internally, the locally unique ID is prefixed with the component ID).
- You need to put the index into the value for `id:scoped` and `:for:scoped`  (there is no special `[]` array suffix for `:scoped`)
- Use the [`key` attribute](http://markojs.com/docs/components/#referencing-nested-html-elements-and-components) if you need to reference an HTML element or a group of HTML elements

I would still like understand why you received an exception with the above stack trace. I'm going to investigate some more.


Does the above solve your problem/answer your question?

**Comment by alexdulin** [comment]
Thanks @patrick-steele-idem that solves it. Appreciate the help.
